File: yaskoam.mrz2.lab1/src/main/scala/yaskoam/mrz2/lab1/matrix/Matrix.scala
package yaskoam.mrz2.lab1.matrix

import scala.collection.parallel.mutable.ParArray

/**
 * @author Q-YAA
 */
class Matrix(private val elements: ParArray[ParArray[Double]]) {

  require(elements.length > 0, "elements array must not be empty")

  def height = elements.length

  def width = elements(0).length

  def get(i: Int, j: Int) = elements(i)(j)

  def set(i: Int, j: Int, value: Double) { elements(i)(j) = value }

  def transpose = new Matrix(ParArray.tabulate(width, height)((i, j) => elements(j)(i)))

  def *(matrix: Matrix) = {
    require(width == matrix.height, "wrong matrices sizes")

    val matrixMultiplicationElements = ParArray.tabulate(height, matrix.width)((i, j) => {
      (for (k <- 0 until matrix.height) yield { get(i, k) * matrix.get(k, j) }).sum
    })

    new Matrix(matrixMultiplicationElements)
  }

  override def toString = (for (row <- elements) yield { row.mkString(" ") }).mkString("\n")
}

File: yaskoam.mrz2.lab1/src/main/java/yaskoam/mrz2/lab1/matrix/Matrix.java
package yaskoam.mrz2.lab1.matrix;

import com.google.common.primitives.Doubles;

/**
 * @author Q-YAA
 */
public class Matrix {

    private double[][] elements;

    private int height;

    private int width;

    public Matrix(double[][] elements) {
        this.elements = elements;
        this.height = elements.length;
        this.width = elements[0].length;
    }

    public int getHeight() {
        return height;
    }

    public int getWidth() {
        return width;
    }

    public double get(int i, int j) {
        return elements[i][j];
    }

    public double[] getRow(int i) {
        return elements[i];
    }

    public double[][] getElements() {
        return elements;
    }

    public Matrix transpose() {

        double[][] transposedElements = new double[getWidth()][getHeight()];

        for (int i = 0; i < width; i++) {
            for (int j = 0; j < height; j++) {
                transposedElements[i][j] = elements[j][i];
            }
        }

        return new Matrix(transposedElements);
    }

    public Matrix multiply(Matrix matrix) {

        double[][] resultElements = new double[height][matrix.getWidth()];

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < matrix.getWidth(); j++) {
                for (int k = 0; k < width; k++) {
                    resultElements[i][j] += elements[i][k] * matrix.get(k, j);
                }
            }
        }

        return new Matrix(resultElements);
    }

    public Matrix multiply(double value) {
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                elements[i][j] *= value;
            }
        }
        return this;
    }

    public Matrix subtract(Matrix matrix) {
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                elements[i][j] -= matrix.get(i, j);
            }
        }
        return this;
    }

    public Matrix add(Matrix matrix) {
        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                elements[i][j] += matrix.get(i, j);
            }
        }
        return this;
    }

    @Override
    public String toString() {
        StringBuilder stringBuilder = new StringBuilder();

        for (double[] row : elements) {
            stringBuilder.append(Doubles.join(" ", row));
            stringBuilder.append("\n");
        }

        return stringBuilder.toString();
    }
}





















